TrafficCapture/captureOffloader/src/main/java/org/opensearch/migrations/trafficcapture/StreamChannelConnectionCaptureSerializer.java

```diff
@@ -197,23 +200,84 @@ private void addStringMessage(int observationFieldNumber, int dataFieldNumber,
         writeByteStringToCurrentStream(dataFieldNumber, str);
     }
 
-    private void addDataMessage(int observationFieldNumber, int dataFieldNumber,
-                                Instant timestamp, ByteBuf buffer) throws IOException {
+    private void addDataMessage(int observationFieldNumber, int dataFieldNumber, Instant timestamp, ByteBuf buffer) throws IOException {
         var byteBuffer = buffer.nioBuffer();
+        int segmentFieldNumber,segmentCountFieldNumber,segmentDataFieldNumber;
+        if (observationFieldNumber == TrafficObservation.READ_FIELD_NUMBER) {
+            segmentFieldNumber = TrafficObservation.READSEGMENT_FIELD_NUMBER;
+            segmentCountFieldNumber = ReadSegment.COUNT_FIELD_NUMBER;
+            segmentDataFieldNumber = ReadSegment.DATA_FIELD_NUMBER;
+        }
+        else {
+            segmentFieldNumber = TrafficObservation.WRITESEGMENT_FIELD_NUMBER;
+            segmentCountFieldNumber = WriteSegment.COUNT_FIELD_NUMBER;
+            segmentDataFieldNumber = WriteSegment.DATA_FIELD_NUMBER;
+        }
+
+        // The required bytes here are not optimizing for space and instead are calculated on the worst case estimate of
+        // the potentially required bytes for simplicity. This could leave ~5 bytes of unused space in the CodedOutputStream
+        // when considering the case of a message that does not need segments or the case of a smaller segment created
+        // from a much larger message
+        int totalMessageBytesRequired = CodedOutputStreamSizeUtil.totalBytesNeededForMessage(timestamp,
+            segmentFieldNumber, segmentDataFieldNumber, segmentCountFieldNumber, 2, byteBuffer, numFlushesSoFar + 1);
+        int totalDataSurroundingBytesRequired = totalMessageBytesRequired - byteBuffer.capacity();
+
+        if (totalDataSurroundingBytesRequired + minDataChunkBytes >= getOrCreateCodedOutputStream().spaceLeft()) {
+            flushCommitAndResetStream(false);
+        }
+
+        if (byteBuffer.limit() == 0 || totalMessageBytesRequired <= getOrCreateCodedOutputStream().spaceLeft()) {
+            addSubstreamMessage(observationFieldNumber, dataFieldNumber, timestamp, byteBuffer);
+            byteBuffer.position(byteBuffer.limit());
+        }
+
+        int dataCount = 0;
+        while(byteBuffer.position() < byteBuffer.limit()) {
+            int availableCOSSpace = getOrCreateCodedOutputStream().spaceLeft();
+            int chunkBytes = totalMessageBytesRequired > availableCOSSpace ? availableCOSSpace - totalDataSurroundingBytesRequired : byteBuffer.limit() - byteBuffer.position();
```

Comment:
Upon working through the arithmetic a couple times, I think that this might be right.  I'd like to see some more exhaustive unit tests that work through a number of edge conditions.  Starting w/ a smaller & buffer size and incrementing it by 1 for a sizeable range might do just that.